doc/design/containerd/port-layer-integration.md

```diff
@@ -0,0 +1,276 @@
+
+*Metadata store*
+
+Metadata is what ties everything together. It will need to link image URIs to digests and links to blobs and snapshots. Given that VIC does not have a persistent filesystem layer, metadata must be persisted into vSphere in some way. 
+
+The question of how this should be done remains open. The KV store on the Port Layer is likely not rich enough to represent the data relationships, so an alternative needs to be sought.
+
+*Events*
+
+At present, the events interface is one of the least clearly defined. However, it’s not hard to imagine how events from vSphere could propagate through containerd via the Port Layer.
+
+**State Management**
+
+One theme that has to be very carefully considered is the question of where state is persisted with containerd integration. Given that there is no persistent filesystem in the VIC appliance and vSphere is the single source of truth for runtime state, it’s important to consider the lifespan of runtime state and where it should be persisted. This does not include caches.
+
+<table>
+  <tr>
+    <td>State</td>
+    <td>Standard containerd/runc</td>
+    <td>VIC containerd</td>
+  </tr>
+  <tr>
+    <td>OCI image metadata</td>
+    <td>Local filesystem</td>
+    <td>Port Layer KV store -> vSphere</td>
+  </tr>
+  <tr>
+    <td>OCI image blob data</td>
+    <td>Local filesystem</td>
+    <td>Local ramfs (not persisted)*</td>
+  </tr>
+  <tr>
+    <td>OCI flat bundle data</td>
+    <td>Local filesystem</td>
+    <td>Single VMDK (support?)</td>
+  </tr>
+  <tr>
+    <td>OCI config data</td>
+    <td>Local filesystem</td>
+    <td>cVM guestInfo</td>
+  </tr>
+  <tr>
+    <td>Image / RW snapshots</td>
+    <td>Btrfs / DeviceMapper</td>
+    <td>VMDK snapshots</td>
+  </tr>
+  <tr>
+    <td>Ctr create/start/stop</td>
+    <td>/proc PID</td>
+    <td>VM exists / powered on / off</td>
+  </tr>
+  <tr>
+    <td>Running container state</td>
+    <td>Local filesystem (via runc)</td>
+    <td>Tether via Port Layer</td>
+  </tr>
+  <tr>
+    <td>Container metrics</td>
+    <td>Host pprof (+ agent?)</td>
+    <td>vSphere via Port Layer</td>
+  </tr>
+  <tr>
+    <td>Container logs</td>
+    <td>Local filesystem</td>
+    <td>File-back serial port</td>
+  </tr>
+  <tr>
+    <td>Stdout/err/in streams</td>
+    <td>Fifo on local filesystem</td>
+    <td>Interaction Port Layer</td>
+  </tr>
+  <tr>
+    <td>Exec’d process state</td>
+    <td>/proc PID ?</td>
+    <td>Tether via Port Layer</td>
+  </tr>
+</table>
+
+
+* There’s a question of whether VIC should support the notion of creating a flat bundle without snapshots. OCI supports this idea by extracting layers on top of each other on the local filesystem. There are potential performance advantages, but these would be significantly offset by the lack of layer reuse.
```

Comment:
For production workloads this is not the concern at all. Since in majority of the cases companies are running specific product for a relatively long time, so the time to merge bunch of layers together to come up with a snapshot is going to be negligible in comparison to everything else. Nevertheless, if there is a way to fake layers on our side, we should consider this approach of layer reuse, since potentially VIC engine may become a tool of developers or QA.